crates/valence_network/src/legacy_ping.rs

```diff
@@ -0,0 +1,305 @@
+        len += self.description.encode_utf16().count();
+
+        len
+    }
+    /// Constructs a new basic [`ServerListLegacyPingResponse`].
+    ///
+    /// See [`description`][Self::description] and [`version`][Self::version].
+    pub fn new(protocol: i32, online_players: i32, max_players: i32) -> Self {
+        Self {
+            protocol,
+            version: String::new(),
+            online_players,
+            max_players,
+            description: String::new(),
+        }
+    }
+    /// Sets the description of the server.
+    ///
+    /// If the resulting response packet is too long to be valid, the
+    /// description will be truncated.
+    ///
+    /// Use [`max_description`][Self::max_description] method to get the max
+    /// valid length for this specific packet with the already set fields
+    /// (version, protocol, online players, max_players).
+    ///
+    /// Also any null bytes will be removed.
+    pub fn description(mut self, description: String) -> Self {
+        self.description = description;
+
+        let overflow = self.length() - Self::MAX_VALID_LENGTH;
+        self.description.truncate(self.description.len() - overflow);
+
+        self.description.retain(|c| c != '\0');
+
+        self
+    }
+    /// Sets the version of the server.
+    ///
+    /// If the resulting response packet is too long to be valid, the
+    /// version will be truncated.
+    ///
+    /// Use [`max_version`][Self::max_version] method to get the max valid
+    /// length for this specific packet with the already set fields
+    /// (description, protocol, online players, max_players).
+    ///
+    /// Also any null bytes will be removed.
+    pub fn version(mut self, version: String) -> Self {
+        self.version = version;
+
+        let overflow = self.length() - Self::MAX_VALID_LENGTH;
+        self.version.truncate(self.version.len() - overflow);
+
+        self.version.retain(|c| c != '\0');
+
+        self
+    }
+    /// Returns the maximum number of characters (not bytes) that this packet's
+    /// description can have with all other fields set as they are.
+    pub fn max_description(&self) -> usize {
+        Self::MAX_VALID_LENGTH - (self.length() - self.description.encode_utf16().count())
+    }
+    /// Returns the maximum number of characters (not bytes) that this packet's
+    /// version can have with all other fields set as they are.
+    pub fn max_version(&self) -> usize {
+        Self::MAX_VALID_LENGTH - (self.length() - self.version.encode_utf16().count())
+    }
+}
+
+// Returns the length of a string representation of a signed integer
+fn int_len(num: i32) -> usize {
+    let num_abs = num.abs() as f64;
+
+    if num < 0 {
+        (num_abs.log10() + 2.0) as usize // because minus sign
+    } else {
+        (num_abs.log10() + 1.0) as usize
+    }
+}
+
+// Removes all `§` and their modifiers, if any
+fn remove_formatting(input: &str) -> String {
```

Comment:
Should there be a function somewhere to remove `\0` in the case where that's used as the separator?

Reply:
`\0` are removed by the constructor of `ServerListLegacyPingResponse` (line [242](https://github.com/valence-rs/valence/pull/406/files/a25542f0b951d95b1c917afb37e557aeb94b72b7#diff-52d9e3c43798d1b5996c379d4a3bd84c436c6b876211a08e482271651b00cb84R242) and [262](https://github.com/valence-rs/valence/pull/406/files/a25542f0b951d95b1c917afb37e557aeb94b72b7#diff-52d9e3c43798d1b5996c379d4a3bd84c436c6b876211a08e482271651b00cb84R262))